File: .me/scripts/org.vifm.FileManager1.rb
#!/bin/env ruby
# https://www.freedesktop.org/wiki/Specifications/file-manager-interface

require 'dbus'

def open_file_manager(uri, select: false)
  args = ['kitty', 'vifm']
  args.append '--select' if select

  path = uri
  path = path[7..] if path.start_with? 'file://'
  args.append path

  system "#{args.join(' ')}&"
end

class FileManager < DBus::Object
  dbus_interface 'org.freedesktop.FileManager1' do
    dbus_method :ShowFolders, 'in uris:as, in startupId:s' do |uris, _|
      open_file_manager uris.first
    end

    dbus_method :ShowItems, 'in uris:as, in startupId:s' do |uris, _|
      open_file_manager uris.first, select: true
    end

    dbus_method :ShowItemProperties, 'in uris:as, in startupId:s' do |_, _|
      puts 'Not Implemented'
    end
  end
end

bus = DBus.session_bus
service = bus.request_service 'org.freedesktop.FileManager1'
fm = FileManager.new '/org/freedesktop/FileManager1'
service.export fm

loop = DBus::Main.new
loop << bus
loop.run

File: .me/scripts/org.vifm.FileManager1.py
#!/usr/bin/env python

import dbus
import dbus.service
import subprocess
import os
from gi.repository import GLib
from dbus.mainloop.glib import DBusGMainLoop

OPATH    = "/org/freedesktop/FileManager1"
IFACE    = "org.freedesktop.FileManager1"
BUS_NAME = "org.freedesktop.FileManager1"

loop = None

def open_filemanager(uri, select=False):
    args = ["kitty", "-e", "vifm"]
    if select: args.append("--select")

    path = str(uri)
    if path.startswith("file://"): path = path[7:]
    args.append(path)

    if os.fork() == 0:
        subprocess.Popen(args)
        os._exit(0)
    else:
        os.wait()

class FileManager1(dbus.service.Object):
    def __init__(self):
        self.loop = GLib.MainLoop()

        bus = dbus.SessionBus()
        bus.request_name(BUS_NAME)
        busname = dbus.service.BusName(BUS_NAME, bus)
        dbus.service.Object.__init__(self, busname, OPATH)

    @dbus.service.method(IFACE, in_signature="ass", out_signature="")
    def ShowFolders(self, uris, _):
        open_filemanager(uris[0])

    @dbus.service.method(IFACE, in_signature="ass", out_signature="")
    def ShowItems(self, uris, _):
        open_filemanager(uris[0], select=True)

    @dbus.service.method(IFACE, in_signature="ass", out_signature="")
    def ShowItemProperties(self, uris, _):
        open_filemanager(uris[0], select=True)

    @dbus.service.method("org.freedesktop.FileManager1",
                         in_signature="", out_signature="")
    def Exit(self):
        self.loop.quit()

if __name__ == "__main__":
    DBusGMainLoop(set_as_default=True)
    fm = FileManager1()
    fm.loop.run()
